### PR TITLE
Added check if particles are not in ELC gap region

### PR DESCRIPTION
**Pull request by biermanncarl**
Fixes Issue #4009

Description of changes:
- add a check in both positive and negative direction with error message

### CODE REVIEW

File: src/core/electrostatics_magnetostatics/elc.cpp [hunk 1]
```diff
@@ -247,6 +247,16 @@ static void add_dipole_force(const ParticleRange &particles) {
   gblcblk[2] = 0; // sum q_i
 
   for (auto const &p : local_particles) {
+    /* check if charged particles are in valid region */
+    if ((p.p.q != 0) && ((p.r.p[2] > elc_params.h) || (p.r.p[2] < 0))) {
+      if (p.r.p[2] < 0)
+        runtimeErrorMsg() << "Particle " << p.p.identity << " entered ELC gap "
```
Comment (jngrad) [inline src/core/electrostatics_magnetostatics/elc.cpp]: We probably want the same check in `double dipole_energy()`, in which case this block can be moved to a free function, e.g. `inline void elc_gap_sanity_check(Particle const &p)`. We also need to add a test to make sure the `runtimeErrorMsg()` is properly converted to a Python exception in the Cython code, e.g. with
```python
    def test_elc_gap_exception(self):
        ... # set up an ELC system, or re-use one that has already been set up in another test
        self.system.constraints.clear() # remove walls, if any was used to minimize the system
        self.system.part[0].pos = [box_l, box_l, box_l + elc_gap - 1]
        with self.assertRaises(Exception):
            self.system.analysis.energy()
        with self.assertRaises(Exception):
            self.integrator.run(2)
```

File: src/core/electrostatics_magnetostatics/elc.cpp [hunk 2]
```diff
@@ -247,6 +247,16 @@ static void add_dipole_force(const ParticleRange &particles) {
   gblcblk[2] = 0; // sum q_i
 
   for (auto const &p : local_particles) {
+    /* check if charged particles are in valid region */
+    if ((p.p.q != 0) && ((p.r.p[2] > elc_params.h) || (p.r.p[2] < 0))) {
```
Comment (KaiSzuttor) [inline src/core/electrostatics_magnetostatics/elc.cpp]: why are you checking for the same condition in the next line?

Comment (biermanncarl) [reply src/core/electrostatics_magnetostatics/elc.cpp]: Well, only part of the same condition. The outer condition checks whether there is an error, at all. The inner condition to distinguish between the two different directions. I could do two independent checks. But then I would have to check ```q!=0``` twice. Or we could leave away the information about how far the particle is in the forbidden region, then I could leave the inner condition away.

Comment (jngrad) [reply src/core/electrostatics_magnetostatics/elc.cpp]: what about checking for the charge in the outer conditional, and using `if(p.r.p[2] > elc_params.h) ... else if(p.r.p[2] < 0))` in the inner conditional?

Comment (KaiSzuttor) [reply src/core/electrostatics_magnetostatics/elc.cpp]: i think this should have a name and should be moved out to a function. the current implementation is happening in the function ```add_dipole_force```.

**Comment by jngrad** [comment]
please use `maintainer/CI/fix_style.sh` :-)

**Comment by biermanncarl** [comment]
> please use `maintainer/CI/fix_style.sh` :-)

I tried, but doesn't work on my local machine properly yet ... It didn't do anything. Sorry.

**Review by jngrad — DISMISSED** [review]
thanks!

**Comment by jngrad** [comment]
Could you please amend these two paragraphs to explain the method will throw if particles enter the gap?
https://github.com/espressomd/espresso/blob/25187188e4271acf6a15493b22bfc526ff72f77d/doc/sphinx/electrostatics.rst#L258-L259
https://github.com/espressomd/espresso/blob/25187188e4271acf6a15493b22bfc526ff72f77d/doc/sphinx/magnetostatics.rst#L99-L100
After that you can remove the WIP from the title, and we're good to merge.

**Review by jngrad — APPROVED** [review]

**Pull request merged**